feat(yp/crypto): BN254 ecc ops

ludamad opened the pull request:
Adds a barrentenberg-backed BN254 ECC ops module to foundation. Currently uses WASM backend, this will change soon as the default backend policy changes to native for node code.

- Added `Bn254G1Point` and `Bn254G2Point` classes using foundation field types (`Fq` for coordinates, `Fr` for scalars)
- G2 extension field coordinates represented as `[Fq, Fq]` tuples
- Point operations (multiplication, curve membership checks) backed by barretenberg

**Barretenberg integration:**
- Added bbapi wrappers for BN254 operations (point multiplication, isOnCurve checks, square root)
- Programmatic curve constant generation via msgpack interface
- Proper type-specific msgpack schemas for serialization

**Cleanup:**
- Don't expose bb.js Fr and delete Fq. For now, bb.js gives you raw buffers and expects you to wrap them. We don't expect much use of bb.js for cryptography work outside of foundation package.

johnathan79717 commented on barretenberg/cpp/src/barretenberg/serialize/msgpack_impl/schema_name.hpp:
```diff
@@ -1,5 +1,9 @@
 #pragma once
 #include "barretenberg/ecc/curves/bn254/g1.hpp"
+#include "barretenberg/ecc/curves/bn254/g2.hpp"
+#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
+#include "barretenberg/ecc/curves/secp256k1/secp256k1.hpp"
+#include "barretenberg/ecc/curves/secp256r1/secp256r1.hpp"
```
Are these used?

ludamad replied:
Yes by the terrible details of argument-dependent lookup. There should be a better way, this is a bit hacky, but needed.

johnathan79717 commented on yarn-project/foundation/src/crypto/index.ts:
```diff
@@ -7,6 +7,7 @@ export * from './poseidon/index.js';
 export * from './secp256k1-signer/index.js';
 export * from './keys/index.js';
 export * from './aes128/index.js';
+export * from './bn254/index.js';
```
Is this used?

ludamad replied:
This is sorta the entire point, so this *can* be used :)

ludamad commented on barretenberg/ts/src/barretenberg_wasm/barretenberg_wasm_main/barretenberg.wasm:
woops - dropping

ludamad commented on yarn-project/foundation/src/fields/fields.ts:
```diff
@@ -321,7 +321,7 @@ export class Fr extends BaseField {
    */
   async sqrt(): Promise<Fr | null> {
-    await BarretenbergSync.initSingleton({ wasmPath: process.env.BB_WASM_PATH });
+    await BarretenbergSync.initSingleton();
```
ignore BB_WASM_PATH stuff, it is going away

ludamad commented on yarn-project/simulator/src/public/avm/opcodes/ec_add.ts:
```diff
@@ -78,14 +78,13 @@ export class EcAdd extends Instruction {
       dest = p2;
     } else if (p2IsInfinite) {
       dest = p1;
     } else {
```
I almost went and made Grumpkin and other curve types all have the same interface and in bb.js but walked it back as yarn-project mostly just implements things as they see fit. at least we can stop creating a do-nothing class instance

ludamad commented on barretenberg/cpp/src/barretenberg/bb/curve_constants.cpp:
```diff
@@ -0,0 +1,97 @@
+namespace bb {
+
+struct CurveConstants {
+    // BN254
```
ty AI

ludamad commented on barretenberg/cpp/src/barretenberg/bbapi/bbapi_ecc.hpp:
```diff
@@ -205,4 +206,107 @@ struct Bn254FrSqrt {
+    static constexpr const char MSGPACK_SCHEMA_NAME[] = "Bn254FqSqrt";
+
+    struct Response {
+        static constexpr const char MSGPACK_SCHEMA_NAME[] = "Bn254FqSqrtResponse";
```
ty AI

ludamad commented on barretenberg/cpp/src/barretenberg/bbapi/bbapi_shared.hpp:
```diff
@@ -152,8 +152,29 @@ struct BBApiRequest {
+ */
+#define BBAPI_ERROR(request, msg)                                                                                      \
+    do {                                                                                                               \
+        (request).error_message = (msg);                                                                               \
```
I knew always passing a request object would help with something :) helps not clog up the return types

ledwards2225 commented on barretenberg/cpp/src/barretenberg/bbapi/bbapi_ecc.cpp:
```diff
@@ -6,19 +6,46 @@
+        BBAPI_ERROR(request, "Input point must be on the curve");
+    }
+    auto result = point * scalar;
+    if (!result.on_curve()) {
```
This is and other similar checks are probably not necessary since they follow from the inputs being on the curve

ledwards2225 reviewed (approved):
LGTM

ludamad committed:
feat(yp/crypto): BN254 ecc ops

Adds a barrentenberg-backed BN254 ECC ops module to foundation. Currently uses WASM backend, this will change soon as the default backend policy changes to native for node code.

- Added `Bn254G1Point` and `Bn254G2Point` classes using foundation field types (`Fq` for coordinates, `Fr` for scalars)
- G2 extension field coordinates represented as `[Fq, Fq]` tuples
- Point operations (multiplication, curve membership checks) backed by barretenberg

**Barretenberg integration:**
- Added bbapi wrappers for BN254 operations (point multiplication, isOnCurve checks, square root)
- Programmatic curve constant generation via msgpack interface
- Proper type-specific msgpack schemas for serialization

**Cleanup:**
- Don't expose bb.js Fr and delete Fq. For now, bb.js gives you raw buffers and expects you to wrap them. We don't expect much use of bb.js for cryptography work outside of foundation package.
